Cleanup/protocols

carlos01sword opened the pull request:
This pull request refactors the entire data and service layer (`NetworkClient`, `BreedsRepository`, `DatabaseService`, etc.) to use the struct-based "Protocol Witness" pattern. This change improves testability and clarity of dependencies throughout the codebase. Additionally, DTOs are standardized and mock data is updated for consistency.

**Refactoring to struct-based "Protocol Witness" pattern:**

* Protocols Removed: All service protocols (`NetworkClientProtocol`, `BreedsRepositoryProtocol`, `BreedsFetching`, `DatabaseServiceProtocol`, `FavoritesServiceProtocol`, `SearchServiceProtocol`) have been deleted.
* Instead of defining functions in a protocol, the struct's is now defined by its closure properties.
* `static func live()` Implementations: The "live" production logic (the code that actually performs network requests or database queries) has been moved into a `static func live()` on each struct.  This provides a single, clear entry point for the "production" version of the service while keeping the struct itself just a simple container of functions.

**DTO and data handling:**

* Introduced `CatBreedDTO` as a standalone struct for API data transfer, replacing previous nested DTOs and updating decoding logic accordingly.
* Updated all usages and mock data to use `CatBreedDTO` instead of the previous `BreedsDataService.CatBreed` type. 

**Testability and composability:**

* All unit tests have been updated. No longer use or need Mock classes (like `MockNetworkClient`).  Instead, mocks are created on the fly by initializing the struct with a specific closure for the test case.

carlos01sword committed:
refactor: update SearchService protocol

carlos01sword committed:
update FavoritesService protocol

carlos01sword committed:
update: BreedsRepository protocol

carlos01sword committed:
update: DatabaseService protocol

carlos01sword committed:
refactor: move live to extension for organization

carlos01sword committed:
update: BreedsApiService protocol

carlos01sword committed:
refactor: fix CatsAppTests to use the new protocols

carlos01sword committed:
refactor: NetworkClient protocol
- Remove NetworkClientProtocol and replace with NetworkClient struct using a closure property (requestData) for network requests.
- Update all usages to inject NetworkClient as a value
- Update Tests

carlos01sword committed:
project update

victorsousabloom commented on CatsApp/Core/Network/NetworkClient.swift:
```diff
@@ -21,51 +21,50 @@ enum NetworkError: Error, CustomStringConvertible {
-protocol NetworkClientProtocol {
-    func request<T: Decodable>(_ endpoint: Endpoint) async throws -> T
+struct NetworkClient {
+    var requestData: (_ endpoint: Endpoint) async throws -> Data
```
since we don't want that other change the requestData
```suggestion
    public internal(set) var requestData: (_ endpoint: Endpoint) async throws -> Data
```
this means that this closure will be public to read but internal to set. Internal is desirable because you update your implementation during ur testes by adding @testable import packages during your tests which allow u to access internal scope...

My explanation might not be super clear 😓 but if u want feel free to call me.

victorsousabloom commented on CatsApp/Data/Repository/BreedsRepository.swift:
```diff
@@ -1,104 +1,72 @@
-        self.database = database
-    }
+struct BreedsRepository {
+    var fetchPage:
```
same here

victorsousabloom commented on CatsApp/Data/Services/DatabaseService.swift:
```diff
@@ -7,127 +7,106 @@
-        context: ModelContext
-    ) async throws
+struct DatabaseService {
+    var saveBreeds: @MainActor ([CatBreedDTO], ModelContext) throws -> [CatBreed]
```
same here

victorsousabloom commented on CatsApp/Data/Services/FavoritesService.swift:
```diff
@@ -8,35 +8,27 @@
-    @MainActor
-    func isFavorite(_ breed: CatBreed, context: ModelContext) throws -> Bool
+struct FavoritesService {
+    var fetchFavorites: @MainActor (ModelContext) throws -> [CatBreed]
```
here as well

victorsousabloom commented on CatsApp/Data/Services/DatabaseService.swift:
```diff
@@ -7,127 +7,106 @@
+                    guard !existingIds.contains(dto.id) else {
+                        return nil
+                    }
+                    return CatBreed(
```
can u make an extension for DTO here just to generate CatBreed from it?
💅  (this means optional change :p)

carlos01sword replied:
Will do 😁

carlos01sword committed:
refactor: Apply public internal(set) to service properties

carlos01sword committed:
refactor: Move CatBreedDTO to its own file and add toModel() extension
- Moves `CatBreedDTO` to its own file under a new `../DTOs` folder.
- Adds a `toModel()` extension to `CatBreedDTO` to encapsulate the logic for converting the DTO to the `CatBreed` model.

mhelenurm commented on CatsApp/Domain/DTOs/CatBreedDTO.swift:
```diff
@@ -0,0 +1,39 @@
+//
+//  Created by Carlos Costa on 21/10/2025.
+//
+
```
Headers are wasted space! 😄 Btw, for how we handle "code credit" in the main project, in `AppDelegate.swift` there's a log of everyone that's contributed to `member-app-ios` (including dates of when they worked)! We made that change to the project at the same time we deleted all these headers a few years ago.

```suggestion
```

mhelenurm replied:
(Also I can see you deleted these headers in most files, so I'm guessing you didn't leave it in intentionally here! Just thought it was a good place to point out where we've stamped our names in the main repo)

carlos01sword committed:
Update CatsApp/Domain/DTOs/CatBreedDTO.swift

Co-authored-by: Mark Helenurm <mhelenurm@gmail.com>

carlos01sword committed:
fix: removed headers
